src/LCTrackClusterAssociation/TrackClusterAssociationAlgorithm.cc

```diff
@@ -113,81 +112,100 @@ StatusCode TrackClusterAssociationAlgorithm::Run()
         float minEnergyDifference(std::numeric_limits<float>::max());
         float minLowEnergyDifference(std::numeric_limits<float>::max());
 
-        // short circuit this loop with a kd-tree search beforehand
-        // iterating over a std::map is expensive, avoid where possible
-        for (unsigned iPseudoLayer = 0; iPseudoLayer <= m_maxSearchLayer; ++iPseudoLayer)
+        LCTrackStates trackStates;
+        const LCTrack* lcTrack = dynamic_cast<const LCTrack*>(pTrack);
+        if( lcTrack )
         {
-            // save the hash key since we may use it a few times
-            auto hash_key = std::make_pair(pTrack, iPseudoLayer);
-            // see if we have a cached search, otherwise do the search and cache
-            auto cached_result = tracks_to_hits.equal_range(hash_key);
-            if (cached_result.first != tracks_to_hits.end())
+            trackStates = lcTrack->GetTrackStates();
+        }
+        else
+        {
+            const TrackState &trackState(pTrack->GetTrackStateAtCalorimeter());
+            trackStates.push_back( trackState );
+        }
+
+        for (auto const& trackState : trackStates)
+        {
+
+            const CartesianVector &trackPosition(trackState.GetPosition());
+
+
+            // short circuit this loop with a kd-tree search beforehand
+            // iterating over a std::map is expensive, avoid where possible
+            for (unsigned iPseudoLayer = 0; iPseudoLayer <= m_maxSearchLayer; ++iPseudoLayer)
             {
-                for (auto iter = cached_result.first; iter != cached_result.second; ++iter )
+                // save the hash key since we may use it a few times
```

Comment:
The diff is quite difficult to parse for this portion of the implementation (mostly provided by our CMS colleagues when they were doing proof-of-principle studies with Pandora), but I had a quick look over your .cc file in full and couldn't see anything awry.

Reply:
The only change here is indentation, but empty lines are still empty so the diff gets messed up